File: src/com/sjsu/cs146/homework/hw9/hw9.java
package com.sjsu.cs146.homework.hw9;

public class hw9 {
    public static void main(String[] args) {
        TreeNode a = new TreeNode(1);
        TreeNode b = new TreeNode(5);
        TreeNode c = new TreeNode(9);
        TreeNode d = new TreeNode(3, a, null);
        TreeNode e = new TreeNode(8, b, c);
        TreeNode f = new TreeNode(4, d, e);
        System.out.println(lowestCommonAncestor(f, a, d).val);
        System.out.println(lowestCommonAncestor(f, a, b).val);
    }
    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(p.val < root.val  && q.val < root.val) return lowestCommonAncestor(root.left, p, q);
        if (p.val > root.val && q.val > root.val) return lowestCommonAncestor(root.right, p, q);
        return root;
    }
}

File: src/com/sjsu/cs146/homework/hw9/hw9.py
class TreeNode(object):
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val < root.val and q.val < root.val: return self.lowestCommonAncestor(root.left , p, q)
        if p.val > root.val and q.val > root.val: return self.lowestCommonAncestor(root.right, p, q)
        return root

solution = Solution()
a = TreeNode(1)
b = TreeNode(5)
c = TreeNode(9)
d = TreeNode(3, a)
e = TreeNode(8, b, c)
f = TreeNode(4, d, e)
print(solution.lowestCommonAncestor(f, a, d).val)
print(solution.lowestCommonAncestor(f, a, b).val)